python/hail/expr/expression.py

```diff
@@ -1240,19 +2346,63 @@ class Int64Expression(NumericExpression):
 
 
 class StringExpression(AtomicExpression):
-    """Expression of type :class:`hail.expr.types.TString`."""
+    """Expression of type :class:`hail.expr.types.TString`.
+
+    >>> s = functions.capture('The quick brown fox')
+    """
     def __getitem__(self, item):
+        """Slice or index into the string.
+
+        Examples
+        --------
+        .. doctest::
+
+            >>> eval_expr(s[:15])
+            'The quick brown'
+
+            >>> eval_expr(s[0])
+            'T'
+
+        Parameters
+        ----------
+        item : slice or :class:`Int32Expression`
+            Slice or character index.
+
+        Returns
+        -------
+        :class:`StringExpression`
+            Substring or character at index `item`.
+        """
         if isinstance(item, slice):
             return self._slice(TString(), item.start, item.stop, item.step)
-        elif isinstance(item, int):
+        elif isinstance(item, int) or isinstance(item, Int32Expression):
             return self._index(TString(), item)
         else:
             raise NotImplementedError()
 
     def __add__(self, other):
-        if not isinstance(other, StringExpression) or isinstance(other, str):
-            raise TypeError("cannot concatenate 'TString' expression and '{}'".format(
-                other._type.__class__ if isinstance(other, Expression) else other.__class__))
+        """Concatenate strings.
+
+        Examples
+        --------
+        .. doctest::
+
+            >>> eval_expr(s + ' jumped over the lazy dog')
+            'The quick brown fox jumped over the lazy dog'
+
+        Parameters
+        ----------
+        other : :class:`StringExpression`
+            String to concatenate.
+
+        Returns
+        -------
+        :class:`StringExpression`
+            Concatenated string.
+        """
+        other = to_expr(other)
```

Comment:
Bugfix (previously if `other` was just a python str, we'd get a NotImplementedError)